### PR TITLE
new feature autoselect - based on Damian's first stab

### PR DESCRIPTION
**Pull request by parmentelat**
proposal for #254

configurable in the `livereveal` metadata with the `auto_select` flag that can be either
* `"none"` (default) - current behaviour
* `"first"` selects first cell in slide
* `"code"` selects first code cell in slide

### CODE REVIEW

**Comment by damianavila** [comment]
Nice! I will try to review it soon but at first glance, it looks good!
Thanks for proposing the PR and for the configurability (I think it is important to have several options in this case).

**Comment by parmentelat** [comment]
I'm rather happy with that second iteration, please make sure to take another look
this second version is admittedly more intrusive, sorry about that;

but on the bright side, this
* allows me to have the selection follow fragments as they show up, which I find super cool, and 
* I was able to remove the dependency to using `window.location.href`, which sounded like a bad thing especially in the context of jupyterlab - where I believe windows.location.href will not be helpful, but maybe I'm wrong here

**Comment by parmentelat** [comment]
this is a screencast of a rise session on a dummy slideshow with `auto_select = 'code'` and `'auto_select_fragment' = true`


screenflick movie 2

**Comment by parmentelat** [comment]
I was hoping to reach a consensus where 
with `auto_select = 'code'` and '`auto_select_fragment' = true` 
could become the default settings ?

**Comment by parmentelat** [comment]
and btw this relates loosely to #270

**Comment by damianavila** [comment]
> could become the default settings ?

I really like this, but let me think a little bit more about it, putting things as default behavior is disruptive for people's workflows, but maybe this is good enough to take that risk... let me think and play with it.

And thanks for you contributions!!

**Comment by parmentelat** [comment]
fair enough
please don't let that hold you from merging, as the default config in the PR as it stands are totally compatible :)

**Comment by damianavila** [comment]
> please don't let that hold you from merging, as the default config in the PR as it stands are totally compatible :)

Yep, sure!

File: rise/static/main.js [hunk 1]
```diff
@@ -562,25 +618,43 @@ function reveal_cell_index(notebook) {
...
+	&& cell.metadata.slideshow.slide_type == 'fragment';
+  }
   var slide_counter = is_slide(notebook.get_cell(0)) ? -1 : 0;
   var subslide_counter = 0;
+  var fragment_counter = 0;    
   var result = null;
 
   notebook.get_cells().forEach(function (cell, index) {
-    if (result)
+    // console.log(`result=${result}, sl=${slide_counter}/${slide} subsl=${subslide_counter}/${subslide} fr=${fragment_counter}/${fragment}`);
```
Comment (damianavila) [inline rise/static/main.js]: This could be probably removed later... but not a blocker for merging.

File: rise/static/main.js [hunk 2]
```diff
@@ -562,25 +618,43 @@ function reveal_cell_index(notebook) {
...
     }
     if ((slide_counter == slide) && (subslide_counter == subslide)) {
-	result = index;
+	    // keep count of fragments but only on current slide
+      if (is_fragment(cell)) {
+	      fragment_counter += 1;
+	    }
+      // we're on the right slide
+      // do we need to worry about fragment
+      var in_fragment = (! focus_fragment ) ? true //
```
Comment (damianavila) [inline rise/static/main.js]: When `in_fragment` could be false here? Can you explain a little bit more about the whole block? Thanks!

File: rise/static/main.js [hunk 3]
```diff
@@ -611,10 +685,40 @@ function revealMode() {
     // select and focus on current cell
     Jupyter.notebook.select(current_cell_index);
     // Need to delay the action a little bit so it actually focus the selected slide
-    setTimeout(function(){ Jupyter.notebook.get_selected_cell().ensure_focused(); }, 1000);
+    setTimeout(function(){ Jupyter.notebook.get_selected_cell().ensure_focused(); }, 500);
   }
 }
 
+/* I'm unsure how to deal with 2 config flags w/ the promise thing,
+ * so I hard-wire this for now
```
Comment (damianavila) [inline rise/static/main.js]: That's ok.

**Comment by damianavila** [comment]
@parmentelat I have verified that things work as expected but I have wrote down one question/clarification, if you can enlighten me, that would be helpful :wink: Thanks.

**Comment by damianavila** [comment]
> @parmentelat I have verified that things work as expected but I have wrote down one question/clarification, if you can enlighten me, that would be helpful 😉 Thanks.

OK, I get it now with a cup of ☕️ ... long day behind :wink:
Still in the next iteration it would nice to add a comment there explaining a little bit more, and fix some indentation mismatches... still not a blocker. Merging now... and sorry for the delay. I will try to follow with the other PRs today/tomorrow.

**Comment by parmentelat** [comment]
Hi Damian

I'm not exactly sure how the PR business is supposed to work at that point, now that you have merged.

I did one single commit on top of my own branch, that I hope clears up all your requests.
just in case, here it is too:
https://github.com/parmentelat/RISE/commits/autoselect

**Comment by damianavila** [comment]
Hi! I did not want to stop the merge anymore since the comments were not blockers for merging. It is great to know you have some more changes clarifying things. I would suggest you to merge master in (or rebase on top of master) and submit a new PR with those changes.
Sorry for the early merge, I did not want to stop us from it anymore since I have delayed the thing a lot. Next time I will wait for your reply. Thanks again for your patience.

**Pull request merged**